### ISSUE TITLE
TypeError after upgrading pypath

### ISSUE THREAD

**Issue by bnymnsen** [saezlab/pypath#203]
Hi,
After upgrading pypath to latest version through pip, I got following error:
```
Traceback (most recent call last):
  File "d:\crossbar\CROssBAR-Data-Process\uniprot data\biocypher_uniprot_process.py", line 7, in <module>
    from pypath.utils import mapping
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 74, in <module>
    import pypath.inputs.uniprot as uniprot_input
  File "C:\anaconda3\lib\site-packages\pypath\inputs\uniprot.py", line 504, in <module>
    field: str | Iterable[str],
TypeError: unsupported operand type(s) for |: 'type' and '_GenericAlias'
```
After googling it, I saw some suggestions to use either ```typing.Union``` or string for type annotation. Maybe these suggestions can be helpful to solve this issue. 

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Python version: 3.9.12
 - Version: 0.14.21

**Comment by deeenes** [comment]
This is a Python 3.9 compatibility issue, Python 3.10 and above are fine with this notation. The solution is to import `__future__.annotations`:

```
from __future__ import annotations
```

I added this in `0.14.22`.

**Comment by ervau** [comment]
Hi @deeenes,

We used to assign '*' to the organism parameter of the uniprot_data function to download bulk data of all organisms before the latest update of uniprot module.

[`organism = taxonomy.ensure_ncbi_tax_id(organism) ` ](https://github.com/saezlab/pypath/blob/8499aa487b4822259d229c7c70405530736bdae4/pypath/inputs/uniprot.py#L541)

With this new line I'm getting the following error because ensure_ncbi_tax_id() returns None when it takes '*' as input. 

```
TypeError                                 Traceback (most recent call last)
Cell In [8], line 1
----> 1 uniprot.uniprot_data_downloader()

File /Volumes/ErvaSP/GitHub/CROssBAR-BioCypher-Migration/bccb/uniprot_adapter.py:150, in Uniprot.uniprot_data_downloader(self)
    148 self.data = {}
    149 for query_key in tqdm(self.attributes):
--> 150     self.data[query_key] = uniprot.uniprot_data(
    151         query_key, self.organism, self.rev
    152     )
    154     # logger.debug(f"{query_key} field is downloaded")
    156 secondary_ids = uniprot.get_uniprot_sec(None)

File /opt/anaconda3/envs/pypath/lib/python3.9/site-packages/pypath/inputs/uniprot.py:555, in uniprot_data(field, organism, reviewed)
    547 get = {
    548     'query': 'organism:%s%s' % (str(organism), rev),
    549     'format': 'tab',
    550     'columns': field_qs,
    551     'compress': 'yes',
    552 }
    554 c = curl.Curl(url, get = get, silent = False, large = True, compr = 'gz')
--> 555 _ = next(c.result)
    558 _id, *variables = zip(*(
    559     line.strip('\n\r').split('\t')
...
   (...)
    568     for f, v in zip(field, variables)
    569 )

TypeError: 'NoneType' object is not an iterator
```

Here is the related part of the log file:

```
[2022-11-11 20:26:06] [curl] Creating Curl object to retrieve data from `https://legacy.uniprot.org/uniprot/`
[2022-11-11 20:26:06] [curl] GET parameters added to the URL: `query=organism%3ANone+AND+reviewed%3A+yes&format=tab&columns=id%2Cid&compress=yes`
[2022-11-11 20:26:06] [curl] Cache file path: `/Users/ervaulusoy/.pypath/cache/e47400efb72210046e4e4baa98ef085a-`
[2022-11-11 20:26:06] [curl] Setting up and calling pycurl.
[2022-11-11 20:26:07] [curl] Empty file retrieved, attempting downlad again
[2022-11-11 20:26:07] [curl] Empty file retrieved, attempting downlad again
[2022-11-11 20:26:07] [curl] Empty file retrieved, attempting downlad again
[2022-11-11 20:26:07] [curl] Download error: HTTP 0
[2022-11-11 20:26:07] [curl] Download error: empty file retrieved.
[2022-11-11 20:26:07] [curl] First 5000 bytes of response:
[2022-11-11 20:26:07] [curl] Download failed, removing the resulted file.
[2022-11-11 20:26:07] [curl] Removing file: `/Users/ervaulusoy/.pypath/cache/e47400efb72210046e4e4baa98ef085a-`
[2022-11-11 20:30:08] [taxonomy] Could not map to NCBI Taxonomy ID: `*`.
```


How should we download uniprot data for all organisms? It will take too long to query all organism ids one by one.

**Comment by deeenes** [comment]
Hi @ervau, You are right, we should allow `"*"` here. I fixed it in 0950ea84 (`0.14.23`).

**Comment by bnymnsen** [comment]
Hi again @deeenes,

After solving the issues above, I am getting the following error:
```
Traceback (most recent call last):
  File "d:\crossbar\CROssBAR-Data-Process\uniprot data\biocypher_uniprot_process.py", line 511, in <module>
    driver.write_nodes(uniprot_adapter.get_uniprot_nodes())
  File "d:\crossbar\CROssBAR-Data-Process\uniprot data\biocypher_uniprot_process.py", line 377, in get_uniprot_nodes
    _props[arg], ensg_ids = self.ensembl_process(_props[arg])
  File "d:\crossbar\CROssBAR-Data-Process\uniprot data\biocypher_uniprot_process.py", line 321, in ensembl_process
    mapping.map_name(
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 3322, in map_name
    return mapper.map_name(
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 1785, in map_name
    mapped_names = self._map_name(
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 2317, in _map_name
    tbl = self.which_table(
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 1484, in which_table
    reader = MapReader(
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 235, in __init__
    self.load()
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 265, in load
    self.read()
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 423, in read
    getattr(self, method)()
  File "C:\anaconda3\lib\site-packages\pypath\utils\mapping.py", line 865, in read_mapping_biomart
    for rec in biomart_data:
  File "C:\anaconda3\lib\site-packages\pypath\inputs\biomart.py", line 136, in biomart_query
    c = curl.Curl(biomart_url, large = True, silent = False)
  File "C:\anaconda3\lib\site-packages\pypath\share\curl.py", line 947, in __init__
    self.curl_setup()
  File "C:\anaconda3\lib\site-packages\pypath\share\curl.py", line 1277, in curl_setup
    self.set_target()
  File "C:\anaconda3\lib\site-packages\pypath\share\curl.py", line 1225, in set_target
    self.target = open(target_path, 'wb')
OSError: [Errno 22] Invalid argument: 'C:\\Users\\ASUS\\.pypath\\cache\\b42ec47e621f0d35d9b1c9e402f618e0-Query>'
```
With using uniprot_adapter script in our migration repo, we are trying to map from ensembl transcript ids to ensembl gene ids using ```map_name``` function of ```mapping.py```. I am getting this error in the mapping part of our uniprot_adapter script. Unfortunately, my log file doesn't contain related information about this mapping error.
